src/wp-includes/rest-api/endpoints/class-wp-rest-attachments-controller.php

```diff
@@ -275,6 +275,17 @@ protected function insert_attachment( $request ) {
 		$attachment->post_mime_type = $type;
 		$attachment->guid           = $url;
 
+		// If the title was not set, use the original filename.
+		if ( empty( $attachment->post_title ) && ! empty( $files['file']['name'] ) ) {
+			// Remove the file extension (after the last `.`)
+			$tmp_title = substr( $files['file']['name'], 0, strrpos( $files['file']['name'], '.' ) );
+
+			if ( ! empty( $tmp_title ) ) {
+				$attachment->post_title = $tmp_title;
+			}
+		}
+
+		// Fall back to the original approach.
```

Comment:
"original" will be ambiguous after time. Maybe there's another way we can phrase this?